R/backend-file.R

```diff
@@ -0,0 +1,507 @@
+
+    keyring_create = function(keyring)
+      b_file_keyring_create(self, private, keyring),
+    keyring_delete = function(keyring = NULL)
+      b_file_keyring_delete(self, private, keyring),
+
+    keyring_lock = function(keyring = NULL)
+      b_file_keyring_lock(self, private, keyring),
+    keyring_unlock = function(keyring = NULL, password = NULL)
+      b_file_keyring_unlock(self, private, keyring, password),
+    keyring_is_locked = function(keyring = NULL)
+      b_file_keyring_is_locked(self, private, keyring),
+
+    keyring_default = function()
+      b_file_keyring_default(self, private),
+    keyring_set_default = function(keyring)
+      b_file_keyring_set_default(self, private, keyring)
+  ),
+
+  private = list(
+    keyring = NULL,
+
+    keyring_create_direct = function(keyring = NULL, password = NULL,
+      nonce = NULL, items = NULL)
+      b_file_keyring_create_direct(self, private, keyring, password, nonce,
+        items),
+
+    keyring_file = function(keyring = NULL, ...)
+      b_file_keyring_file(self, private, keyring, ...),
+    keyring_read_file = function(keyring = NULL)
+      b_file_read_keyring_file(self, private, keyring),
+    keyring_write_file = function(keyring = NULL, nonce = NULL, items = NULL,
+      key = NULL)
+      b_file_write_keyring_file(self, private, keyring, nonce, items, key),
+
+    key_get = function(keyring = NULL)
+      b_file_key_get(self, private, keyring),
+    key_set = function(key = NULL, keyring = NULL)
+      b_file_key_set(self, private, key, keyring),
+    key_unset = function(keyring = NULL)
+      b_file_key_unset(self, private, keyring),
+    key_is_set = function(keyring = NULL)
+      b_file_key_is_set(self, private, keyring),
+
+    keyring_set = function(keyring = NULL, nonce = NULL, check = NULL,
+      items = NULL)
+      b_file_keyring_set(self, private, keyring, nonce, check, items),
+    keyring_get = function(keyring = NULL)
+      b_file_keyring_get(self, private, keyring),
+    nonce_get = function(keyring = NULL)
+      b_file_nonce_get(self, private, keyring),
+    items_get = function(keyring = NULL)
+      b_file_items_get(self, private, keyring),
+    check_get = function(keyring = NULL)
+      b_file_check_get(self, private, keyring)
+  )
+)
+
+b_file_init <- function(self, private, keyring) {
+  self$keyring_set_default(keyring %||% "~/.keyring")
+
+  invisible(self)
+}
+
+b_file_get <- function(self, private, service, username, keyring) {
+
+  if (self$keyring_is_locked(keyring))
+    self$keyring_unlock(keyring)
+
+  all_items <- private$items_get(keyring)
+  item_matches <- sapply(all_items, `[[`, "service_name") %in% service
+
+  if (!is.null(username)) {
+    item_matches <- item_matches &
+                      sapply(all_items, `[[`, "user_name") %in% username
+  }
+
+  if (sum(item_matches) < 1L)
+    b_file_error("cannot get secret",
+                 "The specified item could not be found in the keychain.")
+
+  sapply(
```

Comment:
Can you pls use `vapply()`  instead  of `sapply()`? Everywhere, possibly.

Reply:
done.